File: zig/monte_carlo.zig
const std = @import("std");

const ThreadData = struct {
    samples: usize,
    seed: u32,
    inside: usize,
};

// Linear Congruential Generator - same formula across all languages
fn lcgRandom(seed: *u32) f64 {
    seed.* = seed.* *% 1664525 +% 1013904223;
    return @as(f64, @floatFromInt(seed.* & 0x7FFFFFFF)) / @as(f64, @floatFromInt(0x7FFFFFFF));
}

fn monteCarloWorker(data: *ThreadData) void {
    data.inside = 0;
    
    var i: usize = 0;
    while (i < data.samples) : (i += 1) {
        const x = lcgRandom(&data.seed);
        const y = lcgRandom(&data.seed);
        if (x * x + y * y <= 1.0) {
            data.inside += 1;
        }
    }
}

pub fn monteCarloOperation(total_samples: usize, num_workers: usize) !void {
    const workers = if (num_workers > 0) num_workers else 1;
    
    const allocator = std.heap.page_allocator;
    var threads = try allocator.alloc(std.Thread, workers);
    defer allocator.free(threads);
    
    var thread_data = try allocator.alloc(ThreadData, workers);
    defer allocator.free(thread_data);
    
    const samples_per_worker = total_samples / workers;
    const remainder = total_samples % workers;
    
    for (0..workers) |i| {
        thread_data[i].samples = samples_per_worker;
        if (i == workers - 1) {
            thread_data[i].samples += remainder;
        }
        thread_data[i].seed = @intCast(12345 + i * 67890);  // Consistent seed pattern
        
        threads[i] = try std.Thread.spawn(.{}, monteCarloWorker, .{&thread_data[i]});
    }
    
    var total_inside: usize = 0;
    for (0..workers) |i| {
        threads[i].join();
        total_inside += thread_data[i].inside;
    }
    
    const pi_estimate: f64 = 4.0 * @as(f64, @floatFromInt(total_inside)) / @as(f64, @floatFromInt(total_samples));
    
    std.debug.print("Monte Carlo Pi Estimation\n", .{});
    std.debug.print("Total samples: {}\n", .{total_samples});
    std.debug.print("Points inside circle: {}\n", .{total_inside});
    std.debug.print("Pi estimate: {d:.6}\n", .{pi_estimate});
    std.debug.print("Error: {d:.6}\n", .{3.141592653589793 - pi_estimate});
}

File: rust/src/monte_carlo.rs
use std::thread;

// Linear Congruential Generator - same formula across all languages
fn lcg_random(seed: &mut u32) -> f64 {
    *seed = seed.wrapping_mul(1664525).wrapping_add(1013904223);
    (*seed & 0x7FFFFFFF) as f64 / 0x7FFFFFFF as f64
}

pub fn monte_carlo_operation(total_samples: usize, num_workers: usize) {
    let samples_per_worker = total_samples / num_workers;
    let remainder = total_samples % num_workers;
    
    let mut handles = vec![];
    
    for worker_id in 0..num_workers {
        let samples = if worker_id == num_workers - 1 {
            samples_per_worker + remainder
        } else {
            samples_per_worker
        };
        
        let handle = thread::spawn(move || {
            let mut seed = (12345 + worker_id * 67890) as u32; // Consistent seed pattern
            let mut inside = 0;
            
            for _ in 0..samples {
                let x = lcg_random(&mut seed);
                let y = lcg_random(&mut seed);
                if x * x + y * y <= 1.0 {
                    inside += 1;
                }
            }
            
            inside
        });
        
        handles.push(handle);
    }
    
    let mut total_inside = 0;
    for handle in handles {
        total_inside += handle.join().unwrap();
    }
    
    let pi_estimate = 4.0 * total_inside as f64 / total_samples as f64;
    
    println!("Monte Carlo Pi Estimation");
    println!("Total samples: {}", total_samples);
    println!("Points inside circle: {}", total_inside);
    println!("Pi estimate: {:.6}", pi_estimate);
    println!("Error: {:.6}", std::f64::consts::PI - pi_estimate);
}